client/windows/CAltarWindow.cpp

```diff
@@ -0,0 +1,136 @@
+/*
+ * CAltarWindow.cpp, part of VCMI engine
+ *
+ * Authors: listed in file AUTHORS in main folder
+ *
+ * License: GNU General Public License v2.0 or later
+ * Full text of license available in license.txt file, in main folder
+ *
+ */
+
+#include "StdInc.h"
+#include "CAltarWindow.h"
+
+#include "../gui/CGuiHandler.h"
+#include "../render/Canvas.h"
+#include "../gui/Shortcut.h"
+#include "../widgets/Buttons.h"
+#include "../widgets/TextControls.h"
+
+#include "../CGameInfo.h"
+
+#include "../../lib/CGeneralTextHandler.h"
+#include "../../lib/CHeroHandler.h"
+#include "../../lib/mapObjects/CGHeroInstance.h"
+
+CAltarWindow::CAltarWindow(const IMarket * market, const CGHeroInstance * hero, const std::function<void()> & onWindowClosed, EMarketMode mode)
+	: CWindowObject(PLAYER_COLORED, ImagePath::builtin(mode == EMarketMode::CREATURE_EXP ? "ALTARMON.bmp" : "ALTRART2.bmp"))
+	, hero(hero)
+	, windowClosedCallback(onWindowClosed)
+{
+	OBJECT_CONSTRUCTION_CUSTOM_CAPTURING(255 - DISPOSE);
+
+	if(mode == EMarketMode::ARTIFACT_EXP)
+		createAltarArtifacts(market, hero);
+	else if (mode == EMarketMode::CREATURE_EXP)
+		createAltarCreatures(market, hero);
+	else
+		close();
```

Comment:
How this can happen? Perhaps simply throw runtime error if caller created altar window in unsuitable object?

Reply:
Oooh, closing won't work from the constructor. Yes, throwing an error is a good idea.

Reply:
replaced by assertion